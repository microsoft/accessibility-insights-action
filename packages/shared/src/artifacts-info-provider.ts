// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { injectable } from 'inversify';

@injectable()
export abstract class ArtifactsInfoProvider {
    abstract getArtifactsUrl(): string | undefined;
    abstract getCommitHash(): string | undefined;
}
