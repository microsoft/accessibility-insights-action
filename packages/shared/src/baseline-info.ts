// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { BaselineEvaluation } from 'accessibility-insights-scan';

export type BaselineInfo = {
    baselineFileName?: string;
    baselineEvaluation?: BaselineEvaluation;
};
