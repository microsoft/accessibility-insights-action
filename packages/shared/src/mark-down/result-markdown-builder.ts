// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

import { CombinedReportParameters } from 'accessibility-insights-report';
import { inject, injectable } from 'inversify';
import { ArtifactsInfoProvider } from '../artifacts-info-provider';
import { BaselineInfo } from '../baseline-info';
import { BaselineEvaluation } from 'accessibility-insights-scan';
import { brand } from '../content/strings';
import { bold, escaped, footerSeparator, heading, link, listItem, productTitle, sectionSeparator } from './markdown-formatter';
import { iocTypes } from '../ioc/ioc-types';

@injectable()
export class ResultMarkdownBuilder {
    constructor(@inject(iocTypes.ArtifactsInfoProvider) private readonly artifactsInfoProvider: ArtifactsInfoProvider) {}

    public buildErrorContent(): string {
        const lines = [
            this.headingWithMessage('Something went wrong'),
            sectionSeparator(),
            `You can review the log to troubleshoot the issue. Fix it and re-run the workflow to run the automated accessibility checks again.`,
            sectionSeparator(),
        ];

        return this.scanResultDetails(lines.join(''));
    }

    public buildContent(combinedReportResult: CombinedReportParameters, title?: string, baselineInfo?: BaselineInfo): string {
        const passedChecks = combinedReportResult.results.resultsByRule.passed.length;
        const inapplicableChecks = combinedReportResult.results.resultsByRule.notApplicable.length;
        const failedChecks = combinedReportResult.results.resultsByRule.failed.reduce((a, b) => a + b.failed.length, 0);

        let lines = [
            failedChecks === 0 ? this.headingWithMessage('All applicable checks passed') : this.headingWithMessage(),
            sectionSeparator(),
            this.urlsListItem(
                combinedReportResult.results.urlResults.passedUrls,
                combinedReportResult.results.urlResults.unscannableUrls,
                combinedReportResult.results.urlResults.failedUrls,
            ),
            sectionSeparator(),
            this.rulesListItem(passedChecks, inapplicableChecks, failedChecks),
            sectionSeparator(),

            this.downloadArtifacts(),
            this.failureDetails(combinedReportResult),
        ];

        // baselining is available
        if (baselineInfo !== undefined) {
            lines = [
                this.headingWithMessage(),
                sectionSeparator(),
                this.failureDetailsBaseline(combinedReportResult, baselineInfo),
                sectionSeparator(),
                this.baselineDetails(baselineInfo),
                sectionSeparator(),
                sectionSeparator(),
                this.downloadArtifactsWithLink(combinedReportResult, baselineInfo.baselineEvaluation),
                sectionSeparator(),
                sectionSeparator(),
                footerSeparator(),
                sectionSeparator(),
                heading('Scan summary', 4),
                sectionSeparator(),
                this.urlsListItemBaseline(
                    combinedReportResult.results.urlResults.passedUrls,
                    combinedReportResult.results.urlResults.unscannableUrls,
                    combinedReportResult.results.urlResults.failedUrls,
                ),
                sectionSeparator(),
                this.rulesListItemBaseline(passedChecks, inapplicableChecks, failedChecks),
                sectionSeparator(),
            ];
        }

        if (title !== undefined) {
            lines = [heading(title, 3), sectionSeparator()].concat(lines);
        }

        return this.scanResultDetails(lines.join(''), this.scanResultFooter(combinedReportResult));
    }

    private headingWithMessage = (message?: string): string => {
        if (message) {
            return heading(`${productTitle()}: ${message}`, 3);
        }
        return heading(`${productTitle()}`, 3);
    };

    private baselineDetails = (baselineInfo: BaselineInfo): string => {
        const baselineFileName = baselineInfo.baselineFileName;
        const baselineEvaluation = baselineInfo.baselineEvaluation;
        const baseliningDocsUrl = `https://github.com/microsoft/accessibility-insights-action/blob/main/docs/ado-extension-usage.md#using-a-baseline-file`;
        const baseliningDocsLink = link(baseliningDocsUrl, 'baselining docs');
        const scanArgumentsLink = link(this.artifactsInfoProvider.getArtifactsUrl(), 'scan arguments');
        const baselineNotConfiguredHelpText = `A baseline lets you mark known failures so it's easier to identify new failures as they're introduced. See ${baseliningDocsLink} for more.`;
        const baselineNotDetectedHelpText = `To update the baseline with these changes, copy the updated baseline file to ${scanArgumentsLink}. See ${baseliningDocsLink} for more.`;
        let lines = [''];

        if (baselineFileName === undefined) {
            lines = [bold('Baseline not configured'), sectionSeparator(), baselineNotConfiguredHelpText];
        } else if (baselineEvaluation === undefined) {
            lines = [bold('Baseline not detected'), sectionSeparator(), baselineNotDetectedHelpText];
        } else {
            const newFailures = baselineEvaluation.totalNewViolations;
            const baselineFailures = baselineEvaluation.totalBaselineViolations;
            if (baselineFailures === undefined || (baselineFailures === 0 && newFailures > 0)) {
                lines = [bold('Baseline not detected'), sectionSeparator(), baselineNotDetectedHelpText];
            } else if (baselineFailures > 0) {
                const headingWithBaselineFailures = `${baselineFailures} failure instances in baseline`;
                let baselineFailuresHelpText = `not shown; see ${baseliningDocsLink}`;
                if (newFailures > 0) {
                    baselineFailuresHelpText = baselineFailuresHelpText.concat(' for how to include new failures into the baseline');
                }
                lines = [bold(headingWithBaselineFailures), sectionSeparator(), `(${baselineFailuresHelpText})`];
            }
        }

        return lines.join('');
    };

    private urlsListItem = (passedUrls: number, unscannableUrls: number, failedUrls: number): string => {
        const failedUrlsSummary = `${failedUrls} URL(s) failed, `;
        const passedAndUnscannableUrlsSummary = `${passedUrls} URL(s) passed, and ${unscannableUrls} were not scannable`;
        const urlsSummary = failedUrls === 0 ? passedAndUnscannableUrlsSummary : failedUrlsSummary.concat(passedAndUnscannableUrlsSummary);
        return listItem(`${bold(`URLs`)}: ${urlsSummary}`);
    };

    private urlsListItemBaseline = (passedUrls: number, unscannableUrls: number, failedUrls: number): string => {
        const urlsSummary = `${failedUrls} with failures, ${passedUrls} passed, ${unscannableUrls} not scannable`;
        return `${bold(`URLs`)}: ${urlsSummary}`;
    };

    private rulesListItem = (passedChecks: number, inapplicableChecks: number, failedChecks: number) => {
        const failedRulesSummary = `${failedChecks} check(s) failed, `;
        const passedAndInapplicableRulesSummary = `${passedChecks} check(s) passed, and ${inapplicableChecks} were not applicable`;
        const rulesSummary =
            failedChecks === 0 ? passedAndInapplicableRulesSummary : failedRulesSummary.concat(passedAndInapplicableRulesSummary);
        return listItem(`${bold(`Rules`)}: ${rulesSummary}`);
    };

    private rulesListItemBaseline = (passedChecks: number, inapplicableChecks: number, failedChecks: number) => {
        const rulesSummary = `${failedChecks} with failures, ${passedChecks} passed, ${inapplicableChecks} not applicable`;
        return `${bold(`Rules`)}: ${rulesSummary}`;
    };

    private failureDetails = (combinedReportResult: CombinedReportParameters): string => {
        if (combinedReportResult.results.resultsByRule.failed.length === 0) {
            return '';
        }

        const failedRulesList = combinedReportResult.results.resultsByRule.failed.map((failuresGroup) => {
            const failureCount = failuresGroup.failed.length;
            const ruleId = failuresGroup.failed[0].rule.ruleId;
            const ruleDescription = failuresGroup.failed[0].rule.description;
            return [this.failedRuleListItem(failureCount, ruleId, ruleDescription), sectionSeparator()].join('');
        });
        const lines = [sectionSeparator(), `${heading('Failed instances', 4)}`, sectionSeparator(), ...failedRulesList];

        return lines.join('');
    };

    private failedRuleListItem = (failureCount: number, ruleId: string, description: string) => {
        return listItem(`${bold(`${failureCount} Ã— ${escaped(ruleId)}`)}:  ${escaped(description)}`);
    };

    private failureDetailsBaseline = (combinedReportResult: CombinedReportParameters, baselineInfo: BaselineInfo): string => {
        let lines = [];
        if (this.hasFailures(combinedReportResult, baselineInfo.baselineEvaluation)) {
            const failedRulesList = this.getFailedRulesList(combinedReportResult, baselineInfo.baselineEvaluation);
            const failureInstances = this.getFailureInstances(combinedReportResult, baselineInfo.baselineEvaluation);
            const failureInstancesHeading = this.getFailureInstancesHeading(failureInstances, baselineInfo.baselineEvaluation);
            lines = [sectionSeparator(), bold(failureInstancesHeading), sectionSeparator(), ...failedRulesList];
        } else {
            lines = this.getNoFailuresText(baselineInfo.baselineEvaluation);
        }

        return lines.join('');
    };

    private getNoFailuresText = (baselineEvaluation: BaselineEvaluation): string[] => {
        const checkMark = ':white_check_mark:';
        const pointRight = ':point_right:';
        let failureDetailsHeading = `${checkMark} No failures detected`;
        let failureDetailsDescription = `No failures were detected by automatic scanning.`;
        if (this.baselineHasFailures(baselineEvaluation)) {
            failureDetailsHeading = `${checkMark} No new failures`;
            failureDetailsDescription = 'No failures were detected by automatic scanning except those which exist in the baseline.';
        }
        const nextStepHeading = `${pointRight} Next step:`;
        const tabStopsUrl = `https://accessibilityinsights.io/docs/en/web/getstarted/fastpass/#complete-the-manual-test-for-tab-stops`;
        const tabStopsLink = link(tabStopsUrl, 'Accessibility Insights Tab Stops');
        const nextStepDescription = ` Manually assess keyboard accessibility with ${tabStopsLink}`;

        return [
            bold(failureDetailsHeading),
            sectionSeparator(),
            failureDetailsDescription,
            sectionSeparator(),
            sectionSeparator(),
            bold(nextStepHeading),
            nextStepDescription,
            sectionSeparator(),
        ];
    };

    private getTotalFailureInstancesFromResults = (combinedReportResult: CombinedReportParameters): number => {
        return combinedReportResult.results.resultsByRule.failed.reduce((a, b) => a + b.failed.reduce((c, d) => c + d.urls.length, 0), 0);
    };

    private getFailureInstances = (combinedReportResult: CombinedReportParameters, baselineEvaluation: BaselineEvaluation): number => {
        if (baselineEvaluation) {
            return baselineEvaluation.totalNewViolations;
        }

        return this.getTotalFailureInstancesFromResults(combinedReportResult);
    };

    private getFailureInstancesHeading = (failureInstances: number, baselineEvaluation: BaselineEvaluation): string => {
        let failureInstancesHeading = `${failureInstances} failure instances`;
        if (this.baselineHasFailures(baselineEvaluation)) {
            failureInstancesHeading = failureInstancesHeading.concat(' not in baseline');
        }

        return failureInstancesHeading;
    };

    private getFailedRulesList = (combinedReportResult: CombinedReportParameters, baselineEvaluation: BaselineEvaluation): string[] => {
        if (baselineEvaluation) {
            return this.getNewFailuresList(combinedReportResult, baselineEvaluation);
        }

        return this.getFailedRulesListWithNoBaseline(combinedReportResult);
    };

    private getNewFailuresList = (combinedReportResult: CombinedReportParameters, baselineEvaluation: BaselineEvaluation): string[] => {
        const newFailuresList = [];
        for (const ruleId in baselineEvaluation.newViolationsByRule) {
            const failureCount = baselineEvaluation.newViolationsByRule[ruleId];
            const ruleDescription = this.getRuleDescription(combinedReportResult, ruleId);
            newFailuresList.push([this.failedRuleListItemBaseline(failureCount, ruleId, ruleDescription), sectionSeparator()].join(''));
        }

        return newFailuresList;
    };

    private getFailedRulesListWithNoBaseline = (combinedReportResult: CombinedReportParameters): string[] => {
        const failedRulesList = combinedReportResult.results.resultsByRule.failed.map((failuresGroup) => {
            const failureCount = failuresGroup.failed.reduce((a, b) => a + b.urls.length, 0);
            const ruleId = failuresGroup.failed[0].rule.ruleId;
            const ruleDescription = failuresGroup.failed[0].rule.description;
            return [this.failedRuleListItemBaseline(failureCount, ruleId, ruleDescription), sectionSeparator()].join('');
        });

        return failedRulesList;
    };

    private getRuleDescription = (combinedReportResult: CombinedReportParameters, ruleId: string): string => {
        const matchingFailuresGroup = combinedReportResult.results.resultsByRule.failed.find((failuresGroup) => {
            return failuresGroup.failed[0].rule.ruleId === ruleId;
        });

        return matchingFailuresGroup.failed[0].rule.description;
    };

    private failedRuleListItemBaseline = (failureCount: number, ruleId: string, description: string) => {
        return listItem(`(${failureCount}) ${bold(escaped(ruleId))}:  ${escaped(description)}`);
    };

    private scanResultDetails(scanResult: string, footer?: string): string {
        const lines = [scanResult, sectionSeparator(), footerSeparator(), sectionSeparator(), footer];

        return lines.join('');
    }

    private scanResultFooter(combinedReportResult: CombinedReportParameters): string {
        const axeVersion = combinedReportResult.axeVersion;
        const axeCoreUrl = `https://github.com/dequelabs/axe-core/releases/tag/v${axeVersion}`;
        const axeLink = link(axeCoreUrl, `axe-core ${axeVersion}`);

        return `This scan used ${axeLink} with ${combinedReportResult.userAgent}.`;
    }

    private downloadArtifacts(): string {
        const artifactName = `${brand} artifact`;
        return listItem(`Download the ${bold(artifactName)} to view the detailed results of these checks`);
    }

    private downloadArtifactsWithLink(combinedReportResult: CombinedReportParameters, baselineEvaluation?: BaselineEvaluation): string {
        const artifactsLink = link(this.artifactsInfoProvider.getArtifactsUrl(), 'run artifacts');
        let details = 'all failures and scan details';
        if (!this.baselineHasFailures(baselineEvaluation) && !this.hasFailures(combinedReportResult, baselineEvaluation)) {
            details = 'scan details';
        }
        return `See ${details} by downloading the report from ${artifactsLink}`;
    }

    private baselineHasFailures = (baselineEvaluation: BaselineEvaluation): boolean => {
        return (
            baselineEvaluation !== undefined && baselineEvaluation.totalBaselineViolations && baselineEvaluation.totalBaselineViolations > 0
        );
    };

    private hasFailures = (combinedReportResult: CombinedReportParameters, baselineEvaluation: BaselineEvaluation): boolean => {
        if (baselineEvaluation !== undefined) {
            return baselineEvaluation.totalNewViolations > 0;
        }

        return this.getTotalFailureInstancesFromResults(combinedReportResult) > 0;
    };
}
