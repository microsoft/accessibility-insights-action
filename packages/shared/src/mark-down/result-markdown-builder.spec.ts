// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.
import 'reflect-metadata';

import { ResultMarkdownBuilder } from './result-markdown-builder';
import { CombinedReportParameters } from 'accessibility-insights-report';
import * as axe from 'axe-core';
import * as path from 'path';
import { BaselineEvaluation, CountsByRule } from 'accessibility-insights-scan';
import { BaselineInfo } from '../baseline-info';
import { ArtifactsInfoProvider } from '../artifacts-info-provider';
import { IMock, Mock, MockBehavior, Times } from 'typemoq';

describe(ResultMarkdownBuilder, () => {
    let combinedReportResult: CombinedReportParameters;
    let checkResultMarkdownBuilder: ResultMarkdownBuilder;
    let artifactsInfoProviderMock: IMock<ArtifactsInfoProvider>;

    beforeEach(() => {
        artifactsInfoProviderMock = Mock.ofType<ArtifactsInfoProvider>(undefined, MockBehavior.Strict);
        checkResultMarkdownBuilder = new ResultMarkdownBuilder(artifactsInfoProviderMock.object);
    });

    it('builds error content', () => {
        expect(checkResultMarkdownBuilder.buildErrorContent()).toMatchSnapshot();
    });

    it('builds content with failures', () => {
        setCombinedReportResultWithFailures();

        const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult);

        expect(actualContent).toMatchSnapshot();
    });

    it('builds content with no failures', () => {
        setCombinedReportResultWithNoFailures();

        const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult);

        expect(actualContent).toMatchSnapshot();
    });

    it('escapes content for all possible axe rules', () => {
        const failedRules = axe.getRules().map((r) => ({
            ruleId: r.ruleId,
            description: r.description,
        }));

        combinedReportResult = {
            axeVersion: 'axeVersion',
            userAgent: 'userAgent',
            results: {
                resultsByRule: {
                    failed: failedRules.map((rule) => ({
                        failed: [{ rule }],
                    })),
                    passed: [{}],
                    notApplicable: [{}, {}],
                },
                urlResults: {
                    passedUrls: 1,
                    failedUrls: 1,
                    unscannableUrls: 1,
                },
            },
        } as CombinedReportParameters;

        const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult);

        const snapshotFile = path.join(__dirname, '__custom-snapshots__', 'axe-descriptions.snap.md');
        expect(actualContent).toMatchFile(snapshotFile);
    });

    it('builds content with title', () => {
        const title = 'some title';
        setCombinedReportResultWithNoFailures();

        const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, title);

        expect(actualContent).toMatchSnapshot();
    });

    describe('with baseline', () => {
        let baselineFileName = 'baseline file';

        beforeEach(() => {
            artifactsInfoProviderMock
                .setup((aip) => aip.getArtifactsUrl())
                .returns(() => 'artifacts-url')
                .verifiable(Times.atLeastOnce());
        });

        it('builds content when there are baseline failures and new failures', () => {
            const baselineEvaluation = {
                totalBaselineViolations: 2,
                totalNewViolations: 4,
                newViolationsByRule: { 'rule id': 3, 'rule id 2': 1 } as CountsByRule,
            } as BaselineEvaluation;
            const baselineInfo: BaselineInfo = {
                baselineFileName,
                baselineEvaluation,
            };
            setCombinedReportResultWithFailures();

            const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, undefined, baselineInfo);

            expect(actualContent).toMatchSnapshot();
            verifyAllMocks();
        });

        it('builds content when baseline failures and scanned failures are the same (no new failures)', () => {
            const baselineEvaluation = {
                totalBaselineViolations: 6,
                totalNewViolations: 0,
            } as BaselineEvaluation;
            const baselineInfo: BaselineInfo = {
                baselineFileName,
                baselineEvaluation,
            };
            setCombinedReportResultWithFailures();

            const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, undefined, baselineInfo);

            expect(actualContent).toMatchSnapshot();
            verifyAllMocks();
        });

        it('builds content when there are baseline failures and no additional failures', () => {
            const baselineEvaluation = {
                totalBaselineViolations: 1,
                totalNewViolations: 0,
            } as BaselineEvaluation;
            const baselineInfo: BaselineInfo = {
                baselineFileName,
                baselineEvaluation,
            };
            setCombinedReportResultWithNoFailures();

            const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, undefined, baselineInfo);

            expect(actualContent).toMatchSnapshot();
            verifyAllMocks();
        });

        it('builds content when there are no baseline failures and no new failures', () => {
            const baselineEvaluation = {
                totalBaselineViolations: 0,
            } as BaselineEvaluation;
            const baselineInfo: BaselineInfo = {
                baselineFileName,
                baselineEvaluation,
            };
            setCombinedReportResultWithNoFailures();

            const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, undefined, baselineInfo);

            expect(actualContent).toMatchSnapshot();
            verifyAllMocks();
        });

        it('builds content when there are new failures and no baseline failures', () => {
            const baselineEvaluation = {
                totalBaselineViolations: 0,
                totalNewViolations: 6,
                newViolationsByRule: { 'rule id': 5, 'rule id 2': 1 } as CountsByRule,
            } as BaselineEvaluation;
            const baselineInfo: BaselineInfo = {
                baselineFileName,
                baselineEvaluation,
            };
            setCombinedReportResultWithFailures();

            const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, undefined, baselineInfo);

            expect(actualContent).toMatchSnapshot();
            verifyAllMocks();
        });

        it('builds content when there are no new failures and no baseline detected', () => {
            const baselineEvaluation: BaselineEvaluation = undefined;
            const baselineInfo: BaselineInfo = {
                baselineFileName,
                baselineEvaluation,
            };
            setCombinedReportResultWithNoFailures();

            const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, undefined, baselineInfo);

            expect(actualContent).toMatchSnapshot();
            verifyAllMocks();
        });

        it('builds content when there are new failures and no baseline detected', () => {
            const baselineEvaluation: BaselineEvaluation = undefined;
            const baselineInfo: BaselineInfo = {
                baselineFileName,
                baselineEvaluation,
            };
            setCombinedReportResultWithFailures();

            const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, undefined, baselineInfo);

            expect(actualContent).toMatchSnapshot();
            verifyAllMocks();
        });

        it('builds content when there is no baseline configured', () => {
            baselineFileName = undefined;
            const baselineEvaluation: BaselineEvaluation = undefined;
            const baselineInfo: BaselineInfo = {
                baselineFileName,
                baselineEvaluation,
            };
            setCombinedReportResultWithFailures();

            const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, undefined, baselineInfo);

            expect(actualContent).toMatchSnapshot();
            verifyAllMocks();
        });

        it('builds content when there are no new failures and baseline failures are undefined', () => {
            const baselineEvaluation = {
                totalBaselineViolations: undefined,
            } as BaselineEvaluation;
            const baselineInfo: BaselineInfo = {
                baselineFileName,
                baselineEvaluation,
            };
            setCombinedReportResultWithNoFailures();

            const actualContent = checkResultMarkdownBuilder.buildContent(combinedReportResult, undefined, baselineInfo);

            expect(actualContent).toMatchSnapshot();
            verifyAllMocks();
        });
    });

    const setCombinedReportResultWithFailures = (): void => {
        const ruleInfo1 = { ruleId: 'rule id', description: 'rule description' };
        combinedReportResult = {
            axeVersion: 'axeVersion',
            userAgent: 'userAgent',
            results: {
                resultsByRule: {
                    failed: [
                        {
                            failed: [
                                { rule: ruleInfo1, urls: ['url 1', 'url 2'] },
                                { rule: ruleInfo1, urls: ['url 1'] },
                                { rule: ruleInfo1, urls: ['url 3', 'url 4'] },
                            ],
                        },
                        {
                            failed: [{ rule: { ruleId: 'rule id 2', description: 'rule description 2' }, urls: ['url 5'] }],
                        },
                    ],
                    passed: [{}],
                    notApplicable: [{}, {}],
                },
                urlResults: {
                    passedUrls: 1,
                    failedUrls: 5,
                    unscannableUrls: 7,
                },
            },
        } as CombinedReportParameters;
    };

    const setCombinedReportResultWithNoFailures = (): void => {
        combinedReportResult = {
            axeVersion: 'axeVersion',
            userAgent: 'userAgent',
            results: {
                resultsByRule: {
                    failed: [],
                    passed: [{}],
                    notApplicable: [{}, {}],
                },
                urlResults: {
                    passedUrls: 1,
                    failedUrls: 0,
                    unscannableUrls: 7,
                },
            },
        } as CombinedReportParameters;
    };

    const verifyAllMocks = (): void => {
        artifactsInfoProviderMock.verifyAll();
    };
});
