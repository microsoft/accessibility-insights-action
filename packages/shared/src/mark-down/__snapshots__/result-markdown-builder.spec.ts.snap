// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultMarkdownBuilder builds content with failures 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action
* **URLs**: 5 URL(s) failed, 1 URL(s) passed, and 7 were not scannable
* **Rules**: 4 check(s) failed, 1 check(s) passed, and 2 were not applicable
* Download the **Accessibility Insights artifact** to view the detailed results of these checks
#### Failed instances
* **3 × rule id**:  rule description
* **1 × rule id 2**:  rule description 2

---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder builds content with no failures 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action: All applicable checks passed
* **URLs**: 1 URL(s) passed, and 7 were not scannable
* **Rules**: 1 check(s) passed, and 2 were not applicable
* Download the **Accessibility Insights artifact** to view the detailed results of these checks
---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder builds content with title 1`] = `
"### some title
### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action: All applicable checks passed
* **URLs**: 0 URL(s) passed, and 0 were not scannable
* **Rules**: 0 check(s) passed, and 0 were not applicable
* Download the **Accessibility Insights artifact** to view the detailed results of these checks
---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder builds error content 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action: Something went wrong
You can review the log to troubleshoot the issue. Fix it and re-run the workflow to run the automated accessibility checks again.

---
"
`;

exports[`ResultMarkdownBuilder with baseline builds content when there are baseline failures and new failures 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action

**4 failure instances not in baseline**
* (3) **rule id**:  rule description
* (1) **rule id 2**:  rule description 2

**1 failure instances in baseline**
(not shown; see [baselining docs](temporarily-empty) for how to include new failures into the baseline)

See all failures and scan details by downloading the report from [run artifacts](temporarily-empty)

---
#### Scan summary
**URLs**: 5 with failures, 1 passed, 7 not scannable
**Rules**: 4 with failures, 1 passed, 2 not applicable

---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder with baseline builds content when there are baseline failures and no new failures 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action
**:white_check_mark: No new failures**
No failures were detected by automatic scanning except those which exist in the baseline.

**:point_right: Next step:** Manually assess keyboard accessibility with [Accessibility Insights Tab Stops](https://accessibilityinsights.io/docs/en/web/getstarted/fastpass/#complete-the-manual-test-for-tab-stops)

**1 failure instances in baseline**
(not shown; see [baselining docs](temporarily-empty))

See all failures and scan details by downloading the report from [run artifacts](temporarily-empty)

---
#### Scan summary
**URLs**: 0 with failures, 0 passed, 0 not scannable
**Rules**: 0 with failures, 0 passed, 0 not applicable

---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder with baseline builds content when there are new failures and no baseline detected 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action

**4 failure instances**
* (3) **rule id**:  rule description
* (1) **rule id 2**:  rule description 2

**Baseline not detected**
To update the baseline with these changes, copy the updated baseline file to [scan arguments](temporarily-empty). See [baselining docs](temporarily-empty) for more.

See all failures and scan details by downloading the report from [run artifacts](temporarily-empty)

---
#### Scan summary
**URLs**: 5 with failures, 1 passed, 7 not scannable
**Rules**: 4 with failures, 1 passed, 2 not applicable

---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder with baseline builds content when there are new failures and no baseline failures 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action

**4 failure instances**
* (3) **rule id**:  rule description
* (1) **rule id 2**:  rule description 2

**Baseline not detected**
To update the baseline with these changes, copy the updated baseline file to [scan arguments](temporarily-empty). See [baselining docs](temporarily-empty) for more.

See all failures and scan details by downloading the report from [run artifacts](temporarily-empty)

---
#### Scan summary
**URLs**: 5 with failures, 1 passed, 7 not scannable
**Rules**: 4 with failures, 1 passed, 2 not applicable

---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder with baseline builds content when there are no baseline failures and no new failures 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action
**:white_check_mark: No failures detected**
No failures were detected by automatic scanning.

**:point_right: Next step:** Manually assess keyboard accessibility with [Accessibility Insights Tab Stops](https://accessibilityinsights.io/docs/en/web/getstarted/fastpass/#complete-the-manual-test-for-tab-stops)



See scan details by downloading the report from [run artifacts](temporarily-empty)

---
#### Scan summary
**URLs**: 0 with failures, 1 passed, 7 not scannable
**Rules**: 0 with failures, 1 passed, 2 not applicable

---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder with baseline builds content when there are no new failures and baseline failures are undefined 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action
**:white_check_mark: No failures detected**
No failures were detected by automatic scanning.

**:point_right: Next step:** Manually assess keyboard accessibility with [Accessibility Insights Tab Stops](https://accessibilityinsights.io/docs/en/web/getstarted/fastpass/#complete-the-manual-test-for-tab-stops)

**Baseline not configured**
A baseline lets you mark known failures so it's easier to identify new failures as they're introduced. See [baselining docs](temporarily-empty) for more.

See scan details by downloading the report from [run artifacts](temporarily-empty)

---
#### Scan summary
**URLs**: 0 with failures, 0 passed, 0 not scannable
**Rules**: 0 with failures, 0 passed, 0 not applicable

---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder with baseline builds content when there are no new failures and no baseline detected 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action
**:white_check_mark: No failures detected**
No failures were detected by automatic scanning.

**:point_right: Next step:** Manually assess keyboard accessibility with [Accessibility Insights Tab Stops](https://accessibilityinsights.io/docs/en/web/getstarted/fastpass/#complete-the-manual-test-for-tab-stops)

**Baseline not detected**
To update the baseline with these changes, copy the updated baseline file to [scan arguments](temporarily-empty). See [baselining docs](temporarily-empty) for more.

See all failures and scan details by downloading the report from [run artifacts](temporarily-empty)

---
#### Scan summary
**URLs**: 0 with failures, 0 passed, 0 not scannable
**Rules**: 0 with failures, 0 passed, 0 not applicable

---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;

exports[`ResultMarkdownBuilder with baseline builds content when there is no baseline configured 1`] = `
"### ![Accessibility Insights](https://accessibilityinsights.io/img/a11yinsights-blue.svg) Accessibility Insights Action

**4 failure instances**
* (3) **rule id**:  rule description
* (1) **rule id 2**:  rule description 2

**Baseline not configured**
A baseline lets you mark known failures so it's easier to identify new failures as they're introduced. See [baselining docs](temporarily-empty) for more.

See all failures and scan details by downloading the report from [run artifacts](temporarily-empty)

---
#### Scan summary
**URLs**: 5 with failures, 1 passed, 7 not scannable
**Rules**: 4 with failures, 1 passed, 2 not applicable

---
This scan used [axe-core axeVersion](https://github.com/dequelabs/axe-core/releases/tag/vaxeVersion) with userAgent."
`;
